pkg/pools/pools.go

```diff
@@ -86,7 +90,12 @@ func Copy(dst io.Writer, src io.Reader) (written int64, err error) {
 // NewReadCloserWrapper returns a wrapper which puts the bufio.Reader back
 // into the pool and closes the reader if it's an io.ReadCloser.
 func (bufPool *BufioReaderPool) NewReadCloserWrapper(buf *bufio.Reader, r io.Reader) io.ReadCloser {
+	var closed int32 = 0
 	return ioutils.NewReadCloserWrapper(r, func() error {
+		if !atomic.CompareAndSwapInt32(&closed, 0, 1) {
+			subsequentCloseWarn("ReadCloserWrapper")
+			return nil
```

Comment:
[The behavior of Close after the first call is undefined.](https://pkg.go.dev/io#Closer)

Code which calls Close more than once on an arbitrary io.Closer is buggy by definition. Logging a warning is a great first step, especially for code which may discard the error return, but it really should also return an error to maximize fail-loudly-ness without actually panicking.

Reply:
In this case, I would tend to agree, at least with not logging, nobody will ever see this log message message because there generally wouldn't be a reason to correlate that with anything.
If someone does see it it will have very little meaning to them.

If it is an error maybe we'll see it show up in CI?
In most cases if there is a bug I would imagine it would be something like passing a closer to another function and both the outer and the inner have close in a defer.
Not technically correct since the outer function should be delegating the close to the inner function.
That can also get hairy, what if the inner function returns an error? It would be difficult to tell if the closer was closed or not.

Anyway, I see your point that this would be a bug that we should catch, but also possibly it would just as easily be a non-issue.
Not to mention in most places we don't check error on close.